ZEP000-wf-stream-processing.md

```diff
@@ -0,0 +1,1421 @@
+* removed records:  
+  * `Job:Create`
+  * `Timer:Create`
+  * `Incident:Create`
+  * `WorkflowInstance:Cancel`
+  * `WorkflowInstance:Event_Occurred` if an event (timer, message, error) is triggered for an element
+
+Technically, these changes break the public API of the exporter. However, these records are mainly used for the internal processing, and it is not likely that they're used by a consumer of the stream.    
+
+Beside the records itself, the behavior of the processing changes slightly (i.e. the sequence of records that are written). The processing of a command produces more events and includes more state transitions. Previously, the processing was more fine granular by processing every single event.
+
+TODO: discuss if the changes could be backward-compatible with the current processing (e.g. by starting the processing from a snapshot).
+
+### Impact on Operate and Tasklist
+
+The proposed changes do not make any impact on Operate and Tasklist.
+
+Operate and Tasklist only process records that are considered events in this
+proposal. Only the `MessageStartEventSubscription` records have changed, which
+are not consumed by either Operate and Tasklist. The removed
+`WorkflowInstance:Event_Occurred` is also not consumed by either one. New
+records have no impact, because they weren't consumed yet.
+
+## Testing
+
+<!--
+
+You should describe what is the overall functionality that should be tested.
+
+If you are omitting tests, explain why, and explain the impact if it fails, specifically the worst case scenario.
+
+In each of the sections below, we should already list known cases that need to be tested in the final implementation, and at which level. The initial version here should be a best of effort: it is perfectly acceptable and expected that this section will be amended during implementation.
+
+### Unit
+### Integration
+### E2E
+
+-->
+
+We should verify the following behavior in tests:
+1. we can continue processing after a restart (i.e. the processing of a command produces the same state as replaying its corresponding events)
+2. we can continue processing after an update (i.e. the events of a previous Zeebe version can be replayed)
+3. we can continue processing after an update with behavior changes (i.e. the processor can handle different behaviors/versions, and the replay of events leads to the same state before and after the update)
+
+The effect of the workflow processing (throughput + latency) and the startup time should be measured in different benchmarks. The assumption is that the processing performance and the startup time improve, or are at least on the same level.  
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+<!--
+Why should we *not* do this?
+-->
+
+The processing of a command produces more records compared to the processing of the single event before. Since all records are written in a batch to ensure an atomic operation, it is more likely to exceed the limit of the message size.   
```

Comment:
I do not understand the first sentence, especially the part "processing of the single event". What is single event in this context? Is the language here consistent with the wording used throughout the rest of the document?

Reply:
In the current engine, we process mainly workflow instance events. 

Current: `Element_Completing` produces `Element_Completed`
New proposal: `Complete_Element` produces `Element_Completing` + `Element_Completed` + `SequenceFlow_Taken` + `Activate_Element`

Let's call it "record" here to be more consistent :writing_hand: